akka-http-core/src/main/scala/akka/http/Http.scala

```diff
@@ -173,6 +174,217 @@ class HttpExt(config: Config)(implicit system: ActorSystem) extends akka.actor.E
+    try {
+      val setup = ConnectionPoolSetup(encrypted = false, options, settings, defaultHeaders, log)
+      val effectiveRequest = request.withEffectiveUri(securedConnection = false)
+      val uri = effectiveRequest.uri
+      val gateway = cachedGateway(HostConnectionPoolSetup(uri.authority.host.toString(), uri.effectivePort, setup))
+      val responsePromise = Promise[HttpResponse]()
+      gateway.onComplete(dispatch(request, responsePromise))(system.dispatcher)
+      responsePromise.future
+    } catch {
+      case e: IllegalUriException ⇒ FastFuture.failed(e)
+    }
+
+  /**
+   * Triggers an orderly shutdown of all host connections pools currently maintained by the [[ActorSystem]].
+   * The returned future is completed when all pools that were live at the time of this method call
+   * have completed their shutdown process.
+   *
+   * If existing pool client flows are re-used or new ones materialized concurrently with or after this
+   * method call the respective connection pools will be restarted and not contribute to the returned future.
+   */
+  def shutdownAllConnectionPools(): Future[Unit] = {
+    import scala.collection.JavaConverters._
+    import system.dispatcher
+    val gateways = hostPoolCache.values().asScala
+    system.log.info("Initiating orderly shutdown of all active host connections pools (w/ configured grace period)...")
+    Future.sequence(gateways.map(_.flatMap(_.shutdown()))).map(_ ⇒ ())
+  }
+
+  private def cachedGateway(setup: HostConnectionPoolSetup)(
+    implicit fm: FlowMaterializer): Future[ConnectionPool.Gateway] = {
+    val promise = Promise[ConnectionPool.Gateway]()
+    hostPoolCache.putIfAbsent(setup, promise.future) match {
+      case null ⇒ // we are the first to attempt starting a pool for this setup
+        val gateway =
+          try ConnectionPool.startHostConnectionPool(setup)
+          catch {
+            case NonFatal(e) ⇒
+              hostPoolCache.remove(setup)
+              promise.failure(e)
+              throw e
+          }
+        promise.success(gateway)
+        gateway.whenShuttingDown.onComplete(_ ⇒ hostPoolCache.remove(setup))(system.dispatcher)
+        promise.future
+
+      case future ⇒ future // return cached instance
+    }
+  }
+
+  private def clientFlow[T](settings: HostConnectionPoolSettings)(
+    poolGateway: (HttpRequest, Any) ⇒ Future[ConnectionPool.Gateway])(
+      implicit system: ActorSystem, fm: FlowMaterializer): Flow[(HttpRequest, T), (Try[HttpResponse], T), Any] = {
+    // a connection pool can never have more than pipeliningLimit * maxConnections requests in flight at any point
+    val parallelism = settings.pipeliningLimit * settings.maxConnections
+    Flow[(HttpRequest, T)].mapAsyncUnordered(parallelism, {
+      case (request, userContext) ⇒
+        val responsePromise = Promise[HttpResponse]()
+        poolGateway(request, userContext).onComplete(dispatch(request, responsePromise))(system.dispatcher)
```

Comment:
I'd say yes. Maybe we should switch to that dispatcher everywhere, since `fm.executionContext` is easier to tweak (if required) than `system.dispatcher`.